Libraries/LibEmail/Message.cpp

```diff
@@ -0,0 +1,398 @@
+#include <AK/Optional.h>
+#include <AK/StringBuilder.h>
+#include <LibEmail/Message.h>
+
+namespace LibEmail {
+
+// Some helper functions
+static Optional<Field> parse_message_field(StringView);
+static Optional<SystemFlag> parse_system_flag(StringView);
+static String parse_address(StringView);
+
+Message Message::create_from_imap_data(StringView source)
+{
+    enum class State : char {
+        Start,
+        InCommand,
+        InString,
+        InNumber,
+        InSet,
+        InBodyText
+    };
+
+    dbg() << "Msg before parsing: " << source << '\n';
+
+    Message result;
+    StringBuilder curr_value;
+    State curr_state = State::Start;
+    Vector<char> set_stack;
+    Field curr_field_name;
+    if (!source.is_empty() && source[0] == '(')
+        source = source.substr(1);
+
+    for (size_t i = 0; i < source.length(); ++i) {
+        switch (curr_state) {
+        case State::Start:
+            if (std::isalpha(source[i])) {
+                curr_state = State::InCommand;
+                --i;
+            } else if (curr_field_name == Field::BodyText) {
+                curr_state = State::InBodyText;
+            } else if (source[i] == '"') {
+                curr_state = State::InString;
+            } else if (std::isdigit(source[i])) {
+                curr_state = State::InNumber;
+                --i;
+            } else if (source[i] == '(') {
+                set_stack.append('(');
+                curr_state = State::InSet;
+            }
+            break;
+        case State::InCommand:
+            if (source[i] != ' ') {
+                curr_value.append(source[i]);
+            } else {
```

Comment:
Instead of doing this:
```cpp
if (...) {
    ...
} else {
    ...
}
break;
```
You can do something like this:
```cpp
if (...) {
    ...
    break;
}
...
break;
```